File: app/src/main/java/com/hfdlys/bjtuselfservice/fragment/grade/GradeViewModel.java
package com.hfdlys.bjtuselfservice.fragment.grade;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.hfdlys.bjtuselfservice.StudentAccountManager;

import java.util.ArrayList;
import java.util.List;

public class GradeViewModel extends ViewModel {
    final private MutableLiveData<List<StudentAccountManager.Grade>> gradeList = new MutableLiveData<>();
    StudentAccountManager studentAccountManager = StudentAccountManager.getInstance();
    public LiveData<List<StudentAccountManager.Grade>> getGradeList() {
        return gradeList;
    }
    public void loadGradeList() {
        List<StudentAccountManager.Grade> grades = new ArrayList<>();
        studentAccountManager.getGrade("ln").thenAccept(grade -> {
            grades.addAll(grade);
            gradeList.postValue(grades);
        })
                .exceptionally(throwable -> {
                    if (throwable.toString().equals("Not loginAa")) {
                        studentAccountManager.loginAa().thenAccept(aBoolean -> {
                            if (aBoolean) {
                                loadGradeList();
                            }
                        });
                    } else if (throwable.toString().equals("Not login")) {
                        studentAccountManager.loginAa().thenAccept(aBoolean -> {
                            if (aBoolean) {
                                loadGradeList();
                            }
                        });
                    } else {
                        gradeList.postValue(null);
                    }
                    return null;
                });
        studentAccountManager.getGrade("lr").thenAccept(grade -> {
            grades.addAll(grade);
            gradeList.postValue(grades);
        })
                .exceptionally(throwable -> {
                    if (throwable.toString().equals("Not loginAa")) {
                        studentAccountManager.loginAa().thenAccept(aBoolean -> {
                            if (aBoolean) {
                                loadGradeList();
                            }
                        });
                    } else if (throwable.toString().equals("Not login")) {
                        studentAccountManager.loginAa().thenAccept(aBoolean -> {
                            if (aBoolean) {
                                loadGradeList();
                            }
                        });
                    } else {
                        gradeList.postValue(null);
                    }
                    return null;
                });

    }
    public MutableLiveData<Boolean> getIsAaLogin() {
        return studentAccountManager.getIsAaLogin();
    }
}

File: app/src/main/java/indi/optsimauth/bjtuselfservicecompose/viewmodel/GradeViewModel.kt
package indi.optsimauth.bjtuselfservicecompose.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import indi.optsimauth.bjtuselfservicecompose.StudentAccountManager
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.future.await
import kotlinx.coroutines.launch

class GradeViewModel : ViewModel() {
    private val studentAccountManager = StudentAccountManager.getInstance()

    private var _gradeList =
        MutableStateFlow<MutableList<StudentAccountManager.Grade>>(mutableListOf())
    val gradeList: StateFlow<MutableList<StudentAccountManager.Grade>> = _gradeList.asStateFlow()


    init {
        loadGradeList()
    }

    fun loadGradeList() {
        try {
            studentAccountManager.getGrade("ln").thenAccept {
                _gradeList.value.addAll(it)
                it.forEach {
                    println("ln123 ${it.courseName}")
                }
            }
        } catch (e: Exception) {
            handleLoginError(e)
        }

        try {
            studentAccountManager.getGrade("lr").thenAccept {
                _gradeList.value.addAll(it)
                it.forEach {
                    println("lr123 ${it.courseName}")
                }
            }
        } catch (e: Exception) {
            handleLoginError(e)
        }


    }

    private fun handleLoginError(throwable: Throwable) {
        when (throwable.message) {
            "Not loginAa", "Not login" -> {
                val loginSuccessful = studentAccountManager.loginAa().thenAccept {
                    if (it) {
                        loadGradeList()
                    }
                }
            }

            else -> _gradeList.value = mutableListOf()
        }
    }

//    val isAaLogin: StateFlow<Boolean>
//        get() = _isAaLogin
}

